core/orch_test.go

```diff
@@ -485,29 +486,68 @@ func TestOrchCheckCapacity(t *testing.T) {
 	assert.Nil(o.CheckCapacity(md.ManifestID))
 }
 
-func TestProcessPayment_GivenRecipientError_ReturnsError(t *testing.T) {
+func TestProcessPayment_GivenRecipientError_ReturnsNil(t *testing.T) {
 	n, _ := NewLivepeerNode(nil, "", nil)
 	recipient := new(pm.MockRecipient)
 	n.Recipient = recipient
 	orch := NewOrchestrator(n)
-	recipient.On("ReceiveTicket", mock.Anything, mock.Anything, mock.Anything).Return("", false, errors.New("mock error"))
+
+	recipient.On("ReceiveTicket", mock.Anything, mock.Anything, mock.Anything).Return("", false, nil)
 
 	err := orch.ProcessPayment(defaultPayment(t), ManifestID("some manifest"))
 
-	assert.Contains(t, err.Error(), "mock error")
+	assert := assert.New(t)
+	assert.Nil(err)
+}
+
+func TestProcessPayment_GivenNoSender_ReturnsError(t *testing.T) {
+	n, _ := NewLivepeerNode(nil, "", nil)
+	recipient := new(pm.MockRecipient)
+	n.Recipient = recipient
+	orch := NewOrchestrator(n)
+	recipient.On("ReceiveTicket", mock.Anything, mock.Anything, mock.Anything).Return("some sessionID", false, nil)
+
+	protoPayment := defaultPayment(t)
+
+	protoPayment.Sender = nil
+
+	err := orch.ProcessPayment(protoPayment, ManifestID("some manifest"))
+
+	assert := assert.New(t)
+	assert.Error(err)
+}
+
+func TestProcessPayment_GivenNoTicketParamsInPayment_DoesNotCacheSessionId(t *testing.T) {
+	n, _ := NewLivepeerNode(nil, "", nil)
+	recipient := new(pm.MockRecipient)
+	n.Recipient = recipient
+	orch := NewOrchestrator(n)
+	recipient.On("ReceiveTicket", mock.Anything, mock.Anything, mock.Anything).Return("some sessionID", false, nil)
+
+	var protoPayment net.Payment
+
+	err := orch.ProcessPayment(protoPayment, ManifestID("some manifest"))
+
+	assert := assert.New(t)
+	assert.Error(err)
+	assert.Empty(n.pmSessions)
 }
 
 func TestProcessPayment_GivenWinningTicketAndRecipientError_DoesNotCacheSessionID(t *testing.T) {
 	n, _ := NewLivepeerNode(nil, "", nil)
 	recipient := new(pm.MockRecipient)
 	n.Recipient = recipient
 	orch := NewOrchestrator(n)
-	recipient.On("ReceiveTicket", mock.Anything, mock.Anything, mock.Anything).Return("some sessionID", true, errors.New("mock error"))
+	manifestID := ManifestID("some manifest")
+	sessionID := "some sessionID"
+	recipient.On("ReceiveTicket", mock.Anything, mock.Anything, mock.Anything).Return(sessionID, true, errors.New("mock error"))
 
-	err := orch.ProcessPayment(defaultPayment(t), ManifestID("some manifest"))
+	err := orch.ProcessPayment(defaultPayment(t), manifestID)
+	errorLogsAfter := glog.Stats.Error.Lines()
 
 	assert := assert.New(t)
-	assert.NotNil(err)
+	assert.Nil(err)
+	assert.NotZero(errorLogsAfter)
```

Comment:
If `glog` logged > 0 errors prior to this test (i.e. due to errors logged during other tests in this file), I believe this assertion would always pass? I think `assert.NotZero(errorLogsAfter-errorsLogsBefore)` would be more robust here.

Reply:
Wouldn't the state of `glog.Stats` be wiped out for every test considering they are unit tests? Regardless, I'll follow your advice and adjust to use `after-before`

Reply:
I don't think `glog.Stats` is reset because `glog` is global. I just tried out this code snippet to confirm that the line count is preserved:

```
package logger

import "github.com/golang/glog"

func LogError(msg string) {
	glog.Errorf("ERROR: %v", msg)
}
```

```
package logger

import (
	"fmt"
	"testing"

	"github.com/golang/glog"
)

func TestLogError1(t *testing.T) {
	LogError("foo")
	LogError("foo")

	logCount := glog.Stats.Error.Lines()

	fmt.Printf("%v\n", logCount)
}

func TestLogError2(t *testing.T) {
	logCount := glog.Stats.Error.Lines()

	fmt.Printf("%v\n", logCount)
}
```